File: apps/android/app/src/main/java/ai/openclaw/app/gateway/CloudflareAccessSession.kt
package ai.openclaw.app.gateway

import kotlinx.serialization.Serializable
import kotlinx.serialization.json.Json
import okhttp3.HttpUrl
import java.net.URI
import java.util.Locale

/** Access credentials belong to an exact HTTPS authority, independently from Gateway pairing. */
@ConsistentCopyVisibility
internal data class CloudflareAccessOrigin private constructor(
  val uri: URI,
) {
  fun contains(url: String): Boolean =
    runCatching {
      val resource = URI(url)
      if (resource.rawUserInfo != null || resource.rawFragment != null) return false
      from(URI(resource.scheme, null, resource.host, resource.port, resource.path, null, null).toString()) == this
    }.getOrDefault(false)

  companion object {
    fun from(url: String): CloudflareAccessOrigin {
      val parsed = runCatching { URI(url) }.getOrNull()
      val scheme = parsed?.scheme?.lowercase(Locale.ROOT)
      val host = parsed?.host?.lowercase(Locale.ROOT)
      if (
        url.length > 4096 || parsed == null || scheme !in setOf("https", "wss") || host.isNullOrEmpty() ||
        parsed.rawUserInfo != null || parsed.rawQuery != null || parsed.rawFragment != null ||
        (parsed.port != -1 && parsed.port !in 1..65535)
      ) {
        throw CloudflareAccessException(CloudflareAccessException.Kind.InvalidGateway)
      }
      // OkHttp canonicalizes literal IPv6 hosts before transport; every grant and pin lookup must use that identity.
      val canonicalHost =
        runCatching {
          HttpUrl
            .Builder()
            .scheme("https")
            .host(host)
            .build()
            .host
        }.getOrNull()
          ?: throw CloudflareAccessException(CloudflareAccessException.Kind.InvalidGateway)
      return CloudflareAccessOrigin(URI("https", null, canonicalHost, if (parsed.port == 443) -1 else parsed.port, null, null, null))
    }
  }
}

internal data class CloudflareAccessApplication(
  val origin: CloudflareAccessOrigin,
  val issuer: URI,
  val audience: String,
)

internal class CloudflareAccessSession(
  val application: CloudflareAccessApplication,
  val subject: String,
  val expiresAt: Double,
  private val token: String,
) {
  fun authorizationHeader(
    url: String,
    now: Double = System.currentTimeMillis() / 1000.0,
  ): String? = token.takeIf { application.origin.contains(url) && expiresAt > now }

  fun validate(now: Double = System.currentTimeMillis() / 1000.0) {
    val claims = CloudflareAccessJWT.appClaims(token, application, now)
    if (claims.subject != subject || claims.expiresAt != expiresAt) {
      throw CloudflareAccessException(CloudflareAccessException.Kind.InvalidSession)
    }
  }

  fun encode(): String =
    Json.encodeToString(
      Stored(application.origin.uri.toString(), application.issuer.toString(), application.audience, subject, expiresAt, token),
    )

  override fun toString(): String = "CloudflareAccessSession(<redacted>)"

  @Serializable
  private class Stored(
    val origin: String,
    val issuer: String,
    val audience: String,
    val subject: String,
    val expiresAt: Double,
    val token: String,
  )

  companion object {
    fun decode(value: String): CloudflareAccessSession {
      if (value.length > 65536) throw CloudflareAccessException(CloudflareAccessException.Kind.InvalidSession)
      val stored = Json.decodeFromString<Stored>(value)
      return CloudflareAccessSession(
        CloudflareAccessApplication(CloudflareAccessOrigin.from(stored.origin), URI(stored.issuer), stored.audience),
        stored.subject,
        stored.expiresAt,
        stored.token,
      )
    }
  }
}

internal class CloudflareAccessException(
  val kind: Kind,
) : Exception(kind.description) {
  enum class Kind(
    val description: String,
  ) {
    InvalidGateway("Enter an HTTPS gateway address without credentials, a query, or a fragment."),
    InvalidApplication("This gateway did not provide valid Cloudflare Access sign-in details. Contact its administrator."),
    ConnectionFailed("Could not reach the gateway’s sign-in service. Check your connection and try again."),
    LoginFailed("Browser sign-in did not complete. Check that your account can access this gateway and try again."),
    TimedOut("Browser sign-in timed out. Start sign-in again to continue."),
    InvalidSession("The Cloudflare Access session could not be verified or has expired. Sign in again."),
    StorageFailed("Could not save the sign-in session securely. Unlock this device and try again."),
  }
}

File: apps/ios/Sources/Gateway/CloudflareAccessSession.swift
import Foundation

/// Access grants belong to an HTTPS authority, independently from Gateway pairing credentials.
struct CloudflareAccessOrigin: Hashable, Codable, Sendable {
    let url: URL

    init(_ url: URL) throws {
        guard var components = URLComponents(url: url, resolvingAgainstBaseURL: false),
              let scheme = components.scheme?.lowercased(), ["https", "wss"].contains(scheme),
              let host = components.host?.lowercased(), !host.isEmpty,
              components.user == nil, components.password == nil,
              components.query == nil, components.fragment == nil,
              components.port.map({ (1...65535).contains($0) }) ?? true,
              url.absoluteString.utf8.count <= 4096
        else { throw CloudflareAccessError.invalidGateway }
        components.scheme = "https"
        components.host = host
        components.path = ""
        if components.port == 443 { components.port = nil }
        guard let origin = components.url else { throw CloudflareAccessError.invalidGateway }
        self.url = origin
    }

    init(from decoder: Decoder) throws {
        try self.init(decoder.singleValueContainer().decode(URL.self))
    }

    func encode(to encoder: Encoder) throws {
        var container = encoder.singleValueContainer()
        try container.encode(self.url)
    }

    func contains(_ url: URL) -> Bool {
        // Resource requests may have queries; credentials and fragments never identify an origin.
        guard var components = URLComponents(url: url, resolvingAgainstBaseURL: false),
              components.user == nil, components.password == nil,
              components.fragment == nil
        else { return false }
        components.query = nil
        guard let resource = components.url else { return false }
        return (try? Self(resource)) == self
    }
}

struct CloudflareAccessApplication: Sendable, Equatable {
    let origin: CloudflareAccessOrigin
    let issuer: URL
    let audience: String
}

struct CloudflareAccessSession: Codable, Sendable, CustomStringConvertible, CustomDebugStringConvertible {
    let origin: CloudflareAccessOrigin
    let issuer: URL
    let audience: String
    let subject: String
    let expiresAt: Date
    private let token: String

    init(application: CloudflareAccessApplication, subject: String, token: String, expiresAt: Date) {
        self.origin = application.origin
        self.issuer = application.issuer
        self.audience = application.audience
        self.subject = subject
        self.token = token
        self.expiresAt = expiresAt
    }

    var description: String {
        "CloudflareAccessSession(<redacted>)"
    }

    var debugDescription: String {
        self.description
    }

    func authorizationHeader(for url: URL, now: Date = Date()) -> String? {
        guard self.origin.contains(url), self.expiresAt > now else { return nil }
        return self.token
    }

    func validate(now: Date = Date()) throws {
        let application = CloudflareAccessApplication(
            origin: self.origin,
            issuer: self.issuer,
            audience: self.audience)
        let claims = try CloudflareAccessJWT.appClaims(self.token, application: application, now: now)
        guard claims.sub == self.subject,
              self.expiresAt.timeIntervalSince1970 == claims.exp
        else { throw CloudflareAccessError.invalidSession }
    }
}

enum CloudflareAccessError: Error, LocalizedError {
    case invalidGateway
    case invalidApplication
    case connectionFailed
    case loginFailed
    case timedOut
    case invalidSession
    case storageFailed

    var errorDescription: String? {
        switch self {
        case .invalidGateway:
            "Enter an HTTPS gateway address without credentials, a query, or a fragment."
        case .invalidApplication:
            "This gateway did not provide valid Cloudflare Access sign-in details. Contact its administrator."
        case .connectionFailed:
            "Could not reach the gateway’s sign-in service. Check your connection and try again."
        case .loginFailed:
            "Browser sign-in did not complete. Check that your account can access this gateway and try again."
        case .timedOut:
            "Browser sign-in timed out. Start sign-in again to continue."
        case .invalidSession:
            "The Cloudflare Access session could not be verified or has expired. Sign in again."
        case .storageFailed:
            "Could not save the sign-in session securely. Unlock this device and try again."
        }
    }
}
